streaming/src/main/scala/org/apache/spark/streaming/receiver/ReceiverSupervisor.scala

```diff
@@ -161,12 +163,20 @@ private[streaming] abstract class ReceiverSupervisor(
       stopReceiver("Restarting receiver with delay " + delay + "ms: " + message, error)
       logDebug("Sleeping for " + delay)
       Thread.sleep(delay)
-      logInfo("Starting receiver again")
-      startReceiver()
-      logInfo("Receiver started again")
+      val scheduledLocations = rescheduleReceiver()
+      if (scheduledLocations.isEmpty || scheduledLocations.contains(host)) {
+        logInfo("Starting receiver again")
+        startReceiver()
+        logInfo("Receiver started again")
+      } else {
+        stop("Receiver is scheduled to another executor", None)
+      }
     }(futureExecutionContext)
   }
 
+  /** Reschedule this receiver and return a candidate executor list */
```

Comment:
Using the term `reschedule` is confusing here. I think what you want is essentially to get a set of acceptable locations from the driver. Wouldne a better name be `getAllowedLocations`?